contracts/external/callees/Trade0x.sol

```diff
@@ -0,0 +1,168 @@
+import {SafeERC20} from "../../packages/oz/SafeERC20.sol";
+import {ERC20Interface} from "../../interfaces/ERC20Interface.sol";
+import {WETH9} from "../canonical-weth/WETH9.sol";
+import {ERC20PermitUpgradeable} from "../../packages/oz/upgradeability/erc20-permit/ERC20PermitUpgradeable.sol";
+
+/**
+ * @author Opyn Team
+ * @title Trade0x
+ * @notice callee contract to trade on 0x.
+ */
+contract Trade0x is CalleeInterface {
+    using SafeERC20 for ERC20Interface;
+
+    ///@dev 0x portocal fee to fill 1 order
+    uint256 private PORTOCAL_FEE_BASE = 70000;
+
+    ZeroXExchangeInterface public exchange;
+    ERC20Interface public weth;
+
+    address public controller;
+    address public assetProxy;
+    address public staking;
+
+    constructor(
+        address _exchange,
+        address _assetProxy,
+        address _weth,
+        address _staking,
+        address _controller
+    ) public {
+        exchange = ZeroXExchangeInterface(_exchange);
+        assetProxy = _assetProxy;
+        weth = ERC20Interface(_weth);
+        weth.safeApprove(_assetProxy, uint256(-1));
+        weth.safeApprove(_staking, uint256(-1));
+        controller = _controller;
+    }
+
+    event Trade(
+        address indexed taker,
+        address indexed maker,
+        address takerAsset,
+        address makerAsset,
+        uint256 takerAmount,
+        uint256 makerAmount
+    );
+
+    /**
+     * @notice fill 0x order
+     * @dev it is dangerous to do an unlimited approval to this contract
+     * @param _sender the original sender who wants to trade on 0x
+     * @param _data abi-encoded order, fillamount, signature and _sender. fee payer is the address we pull weth from.
+     */
+    function callFunction(address payable _sender, bytes memory _data) external override {
+        require(msg.sender == controller, "Trade0x: sender not controller");
+
+        _directlyTrade(_sender, _data);
+    }
+
+    function _directlyTrade(address payable _sender, bytes memory _data) internal {
+        (
+            address trader,
+            ZeroXExchangeInterface.Order[] memory order,
+            uint256[] memory takerAssetFillAmount,
+            bytes[] memory signature
+        ) = abi.decode(_data, (address, ZeroXExchangeInterface.Order[], uint256[], bytes[]));
+
+        require(
+            tx.origin == trader,
+            "Trade0x: funds can only be transferred in from the person sending the transaction"
+        );
+
+        for (uint256 i = 0; i < order.length; i++) {
+            address takerAsset = decodeERC20Asset(order[i].takerAssetData);
+            // transfer takerAsset from trader to this contract
+            ERC20Interface(takerAsset).safeTransferFrom(trader, address(this), takerAssetFillAmount[i]);
+            // approe the 0x ERC20 Proxy to transfer takerAsset from this contract
+            ERC20Interface(takerAsset).safeIncreaseAllowance(assetProxy, takerAssetFillAmount[i]);
```

Comment:
one thing to note here is that if the taker asset is weth this will fail. We are not using weth as taker asset though right? So maybe we should document that? @antoncoding

Reply:
In case `takerAsset` is WETH, why it will fail ?

Reply:
Oh I had this issue yesterday where because in the constructor we do a big allowance for WETH, when we call this again, it runs into a safe math overflow issue.

Reply:
IMO okay to leave it this way since we're unlikely to enable weth / oToken orders in the short term. But we should def add that to comments.